### ISSUE TITLE
Enriched verification criteria lose their shape, so the VLM checklist reads "1. undefined" in 58% of evaluations

### ISSUE THREAD

**Issue by kreuzhofer** [kreuzhofer/chat3d-app#33]
## Summary

Spec generation produces verification criteria as `AnnotatedCriterion[]` — `{ text, visibility }`. Spec **enrichment** returns them as plain `string[]`. The codegen orchestrator overwrites the annotated array with the plain one, and everything downstream still reads `.text`, which is `undefined` on a string.

The visual judge is therefore handed a checklist of literal placeholders:

```
Verification Checklist — answer each with pass, fail, or uncertain:
1. undefined
2. undefined
3. undefined
4. undefined
5. undefined
6. undefined
```

**1,873 of 3,228 stored VLM prompts (58%) contain this.** Of the 2,925 prompts carrying a checklist block at all, **64% are entirely `undefined`**.

## Scale, by category

Percentage of evaluations whose prompt carried a checklist with **real** items:

| category | no checklist | all `undefined` | real items |
| --- | --- | --- | --- |
| Gridfinity Storage | 21% | 79% | **0%** |
| Hinges | 44% | 56% | **0%** |
| PCB Cases | 63% | 36% | **1%** |
| Simple Everyday Objects | 8% | 91% | **1%** |
| Mechanical Components | 6% | 92% | **2%** |
| Primitives | 5% | 93% | **2%** |
| Boolean Operations | 6% | 91% | **3%** |
| Electronic Components | 4% | 92% | **3%** |
| Surface Modifications | 10% | 86% | **4%** |
| Sketch Operations | 1% | 94% | **5%** |
| Extrusions and Revolutions | 4% | 91% | **5%** |
| Arrays and Patterns | 1% | 95% | **5%** |
| Generic Enclosures | 13% | 81% | **6%** |
| bd_warehouse Examples | 6% | 87% | **7%** |
| Missing Examples | 9% | 24% | **67%** |

Every production category runs at **0–7% working checklists**. Only *Missing Examples* is largely intact, and that is the category whose examples skip enrichment.

Note this is not the "hard categories are starved" story it first looks like: the easy categories are worse on the `undefined` axis. Two distinct failures overlap — a missing checklist block, and a present-but-placeholder one — and together they leave nearly the whole corpus without a usable checklist.

## Why it matters

The checklist is the judge's only prompt-specific instruction. Without it the judge falls back on a generic four-band rubric and its own reading of the render, which is exactly the condition under which false positives appear: a model with a missing standoff, a non-protruding stacking lip or a through-hole where a recess was specified scores 8–9 because nothing asked the judge to check.

When the checklist does work it works well: of 251 failed checklist items corpus-wide, only 6 co-occur with a visual score ≥ 8.

Everything measured about judge quality to date — including a Sonnet-vs-local-model comparison showing a 22× difference in false-negative rate — was measured on a starved instrument. Those results are not wrong, but they are not measurements of the intended system.

## Root cause

- The enrichment result type declares `verificationCriteria: string[]`, and its parser filters the LLM response down to strings.
- The codegen orchestrator spreads the enrichment result over the spec result, replacing an `AnnotatedCriterion[]` with a `string[]`.
- Downstream consumers map `.text` to build the checklist, yielding `undefined` per item.

**TypeScript has been reporting this all along.** Three of the repo's 57 known-baseline type errors are precisely this defect — two in the enrichment service and one at the orchestrator's overwrite. The signal existed and was lost inside an accepted error baseline (see #29 for the enforcement gap that allows a baseline this large).

## Desired behavior

Enrichment preserves the annotated shape, so a criterion keeps its visibility annotation through the enrichment step and the checklist reaching the judge contains real text. A criterion that cannot be rendered as text never reaches the judge as a placeholder — if the list is empty the checklist block is omitted rather than emitted with filler.

## Acceptance criteria

- [ ] Enriched criteria retain `text` and `visibility`; the orchestrator's overwrite is type-correct.
- [ ] No prompt sent to the visual judge can contain a placeholder checklist item. An empty or unusable criteria list omits the block entirely.
- [ ] A test asserts that a spec run through enrichment yields a checklist whose items are the criteria texts, not `undefined`.
- [ ] A test asserts the empty case omits the checklist block rather than emitting numbered filler.
- [ ] The three related type errors are gone, and the backend baseline drops from 57 to 54 rather than staying level.
- [ ] The suite shows no new failures.

## Backfill — needs a decision, not assumed here

Roughly 1,873 stored examples were evaluated with a placeholder checklist, and their `visual_score`, `eval_score` and `approval_status` all derive from that evaluation. Options, with the trade-off rather than a recommendation:

- **Re-evaluate all affected examples.** Most correct, most expensive, and it will move approval rates — probably downward in the hard categories, which is the point.
- **Re-evaluate a stratified sample** to size the error, then decide.
- **Leave scores, mark provenance.** Cheapest; keeps the corpus internally inconsistent between rows scored with and without a working checklist.

Whichever is chosen, examples scored before the fix and after are not comparable, and anything using the corpus as a reference set (judge benchmarking, fine-tuning data) needs to know which side of the fix each row sits on.

## Out of scope

- Redesigning the rubric or the judge's output shape (tracked separately).
- The missing-checklist-block case where enrichment never ran, beyond ensuring no placeholder is emitted.
- Re-running any evaluation as part of this change; the backfill is a separate decision above.

**Comment by kreuzhofer** [comment]
> *This was generated by AI during triage.*

## Agent Brief

**Category:** bug
**Summary:** Enriched verification criteria lose their annotated shape, so the visual judge is handed a checklist of literal `undefined` placeholders in the majority of evaluations.

**Current behavior:**

Spec generation produces verification criteria as annotated objects — each carrying the criterion text plus a visibility annotation that routes it to visual or code evaluation. Spec **enrichment** returns criteria as plain strings. The codegen orchestrator replaces the annotated array with the plain one, and every downstream consumer still reads the text property, which does not exist on a string. The judge receives:

```
Verification Checklist — answer each with pass, fail, or uncertain:
1. undefined
2. undefined
3. undefined
4. undefined
5. undefined
6. undefined
```

Verified against the stored prompts: **1,873 of 3,228 (58%)** contain this. Of the 2,925 prompts carrying a checklist block at all, **64% are entirely placeholders**. Measured per category, the share of evaluations whose checklist held real items is **0–7% across every production category**; only the one category that skips enrichment is largely intact, at 67%.

Two distinct failures overlap and should not be conflated: a checklist block that is absent entirely, and one that is present but filled with placeholders. This issue is about the second; the first matters only insofar as an empty list must not be rendered as filler.

The compiler already reports this. Three of the repository's known-baseline type errors are exactly this defect — two where enrichment declares the plain-string shape, one at the orchestrator's overwrite. The signal existed and was lost inside an accepted error baseline.

Note also that several consumers read these criteria back from persisted rows using an unchecked type assertion rather than a validated conversion. Those assertions carry the same defect into the re-render path while hiding it from the compiler, and are part of why this survived.

**Desired behavior:**

A criterion keeps its text and its visibility annotation through enrichment, so the checklist reaching the judge contains the real questions. A criterion that cannot be rendered as text never reaches the judge as a placeholder: when the usable list is empty, the checklist block is omitted rather than emitted with numbered filler.

Where criteria are read back from persistence, the shape is validated rather than asserted, so a stored row in the old shape is detected instead of silently producing placeholders again.

**Key interfaces:**

- The enrichment result's verification-criteria field. It currently declares a plain-string collection; it should carry the same annotated shape the spec generator produces, and the enrichment parser should preserve the annotation rather than filtering down to strings. Where the enrichment model returns bare strings, they need an explicit, deliberate visibility default rather than an implicit one.
- The point in the codegen orchestrator where the enrichment result is merged over the spec result. This is where the shapes collide today.
- The checklist-building step that maps criteria to prompt lines. It should be incapable of emitting a placeholder: a criterion that yields no text is dropped, and a fully empty list suppresses the block.
- The persisted-criteria read paths that currently use unchecked type assertions. These should validate.

**Acceptance criteria:**

- [ ] Enriched criteria retain both their text and their visibility annotation; the orchestrator's merge is type-correct without an assertion.
- [ ] No prompt sent to the visual judge can contain a placeholder checklist item, for any input.
- [ ] An empty or unusable criteria list omits the checklist block entirely rather than emitting numbered filler.
- [ ] A test drives a spec through enrichment and asserts the resulting checklist lines are the criterion texts, not placeholders.
- [ ] A test asserts the empty case omits the block.
- [ ] A test asserts that criteria persisted in the old plain-string shape are handled without producing placeholders.
- [ ] The three related type errors are gone: the backend typecheck baseline drops from 57 to 54 rather than holding level.
- [ ] The suite shows no new failures. Note both packages carry a known non-zero baseline; compare against it rather than expecting zero.

**Out of scope:**

- Re-evaluating or backfilling any stored example. The corpus contains roughly 1,210 approved examples scored against a placeholder checklist, and correcting them changes approval rates and the contents of the training set. That is a maintainer decision and will be tracked separately. Useful context for whoever picks it up: the source data is intact — 2,485 of 2,855 prompt records still hold both the plain checklist and the annotated criteria, and the plain checklist is already phrased as visual questions. A backfill therefore needs only a re-run of visual evaluation, not spec regeneration.
- Redesigning the scoring rubric or the judge's output shape.
- The case where enrichment never ran and no checklist block is emitted at all, beyond ensuring no placeholder is produced.
- The broader typecheck-baseline enforcement gap that allowed three live errors to hide.
